src/posix/routing_manager/routing_manager.cpp

```diff
@@ -0,0 +1,639 @@
+                          mInfraNetif.GetName());
+            mInfraNetif.AddGatewayAddress(newOnLinkPrefix);
+        }
+    }
+    else
+    {
+        if (IsValidOnLinkPrefix(mAdvertisedOnLinkPrefix))
+        {
+            otLogInfoPlat("stop advertising prefix %s on interface %s",
+                          Ip6PrefixString(mAdvertisedOnLinkPrefix).AsCString(), mInfraNetif.GetName());
+            mInfraNetif.RemoveGatewayAddress(mAdvertisedOnLinkPrefix);
+        }
+    }
+
+    if (newOmrPrefix == mLocalOmrPrefix)
+    {
+        if (!IsValidOmrPrefix(mAdvertisedOmrPrefix))
+        {
+            otLogInfoPlat("publish new OMR prefix in Thread network");
+            PublishOmrPrefix(newOmrPrefix);
+        }
+    }
+    else
+    {
+        if (IsValidOmrPrefix(mAdvertisedOmrPrefix))
+        {
+            otLogInfoPlat("there is already OMR prefix in the Thread network, stop publishing");
+            UnpublishOmrPrefix(mAdvertisedOmrPrefix);
+        }
+    }
+
+    if (IsValidOnLinkPrefix(newOnLinkPrefix) || newOmrPrefix != mAdvertisedOmrPrefix)
+    {
+        SendRouterAdvertisement(newOmrPrefix, newOnLinkPrefix);
+    }
+
+    mAdvertisedOnLinkPrefix = newOnLinkPrefix;
+    mAdvertisedOmrPrefix    = newOmrPrefix;
+}
+
+void RoutingManager::SendRouterSolicit()
+{
+    uint32_t timeout;
+
+    mRouterAdvertiser.SendRouterSolicit(mInfraNetif, kKinkLocalAllRouters);
+
+    // We wait a bit longer than the solicition interval.
+    timeout = kRtrSolicitionInterval * 1000 + GenerateRandomNumber(0, 1000);
+    mRouterSolicitTimer.Start(timeout);
+
+    otLogInfoPlat("Router Solicit timer scheduled in %.1f s", timeout / 1000.0f);
+}
+
+void RoutingManager::SendRouterAdvertisement(const otIp6Prefix &aOmrPrefix, const otIp6Prefix &aOnLinkPrefix)
+{
+    uint32_t           nextSendTime;
+    const otIp6Prefix *omrPrefix    = IsValidOmrPrefix(aOmrPrefix) ? &aOmrPrefix : nullptr;
+    const otIp6Prefix *onLinkPrefix = IsValidOnLinkPrefix(aOnLinkPrefix) ? &aOnLinkPrefix : nullptr;
+
+    VerifyOrExit(omrPrefix != nullptr || onLinkPrefix != nullptr);
+
+    mRouterAdvertiser.SendRouterAdvertisement(omrPrefix, onLinkPrefix, mInfraNetif, kLinkLocalAllNodes);
+
+    ++mRouterAdvertisementCount;
+
+    nextSendTime = GenerateRandomNumber(kMinRtrAdvInterval, kMaxRtrAdvInterval);
+
+    if (mRouterAdvertisementCount <= kMaxInitRtrAdvertisements && nextSendTime > kMaxInitRtrAdvInterval)
+    {
+        nextSendTime = kMaxInitRtrAdvInterval;
+    }
+
+    otLogInfoPlat("Router Advertisement scheduled in %u Seconds", nextSendTime);
+    mRouterAdvertisementTimer.Start(nextSendTime * 1000);
+
+exit:
+    return;
+}
+
+bool RoutingManager::IsValidOmrPrefix(const otIp6Prefix &aPrefix)
+{
+    bool isValid = false;
+
+    VerifyOrExit(aPrefix.mLength == OT_IP6_PREFIX_BITSIZE);
+    VerifyOrExit(aPrefix.mPrefix.mFields.m8[0] == 0xfd);
+
+    isValid = true;
+
+exit:
+    return isValid;
```

Comment:
```suggestion
    return (aPrefix.mLength == OT_IP6_PREFIX_BITSIZE) && (aPrefix.mPrefix.mFields.m8[0] == 0xfd);
```

btw, should we allow global OMR like 2001:://64 ?

Reply:
I think duckhorn Border Router is specific to bi-directional connectivity in local network.

Reply:
Seems to fine to assume it's ULA.